Source/Microsoft.Gateway/Microsoft.InnerEye.Listener.Common/Providers/BaseConfigProvider.cs

```diff
@@ -25,110 +25,201 @@ public class BaseConfigProvider<T>
+                    var (t, loaded, parsed) = LoadFile(file);
+                    if (!loaded)
+                    {
+                        // File still in use, FileWatcher has reported file changed but
+                        // the other process has not finished yet.
+                        return false;
+                    }
+
+                    if (parsed)
                     {
                         ts.Add(t);
                     }
                 }
 
-                _ts = ts.ToArray();
+                Config = _flatMap.Invoke(ts);
+
+                return true;
             }
             else
             {
                 var logEntry = LogEntry.Create(ServiceStatus.NewConfigurationError,
                     string.Format("Settings is neither a file nor a folder: {0}", _settingsFileOrFolderName));
                 logEntry.Log(_logger, LogLevel.Error);
+
+                return false;
             }
         }
 
         /// <summary>
         /// Update settings file, according to an update callback function.
         /// </summary>
         /// <param name="updater">Callback to update the settings. Return new settings for update, or the same object to not update.</param>
-        /// <param name="equalityComparer">How to compare objects.</param>
-        protected void UpdateFile(Func<T, T> updater, IEqualityComparer<T> equalityComparer)
+        /// <param name="equalityComparer">Optional, how to compare objects.</param>
+        public (T, bool) Update(Func<T, T> updater, IEqualityComparer<T> equalityComparer = null)
         {
             if (!File.Exists(_settingsFileOrFolderName))
             {
                 throw new NotImplementedException(string.Format("Can only update single settings files: {0}", _settingsFileOrFolderName));
             }
 
-            var (t, loaded) = LoadFile(_settingsFileOrFolderName);
-            if (!loaded)
+            var (t, loaded, parsed) = LoadFile(_settingsFileOrFolderName);
+            if (!loaded || !parsed)
             {
-                return;
+                return (default(T), false);
             }
 
             var newt = updater.Invoke(t);
+
+            equalityComparer = equalityComparer ?? EqualityComparer<T>.Default;
+
             if (equalityComparer.Equals(newt, t))
             {
-                return;
+                return (default(T), false);
             }
 
             SaveFile(newt, _settingsFileOrFolderName);
+
+            return (newt, true);
         }
 
         /// <summary>
         /// Load T from a JSON file.
         /// </summary>
         /// <param name="path">Path to file.</param>
-        /// <returns>Pair of T and true if file loaded correctly, false otherwise.</returns>
-        private (T, bool) LoadFile(string path)
+        /// <returns>Triple of T, true if file loaded correctly, false otherwise, true if file parsed correctly, false otherwise.</returns>
+        private (T, bool, bool) LoadFile(string path)
```

Comment:
Create a small class with names. I dont know what this booleans mean

Reply:
Done